crates/core/src/live_socket/navigation/ffi.rs

```diff
@@ -0,0 +1,241 @@
+//! # FFI Navigation Types
+//!
+//! Types and utilities for interacting with the navigation API for the FFI api consumers.
+use reqwest::Url;
+
+pub type HistoryId = u64;
+
+#[uniffi::export(callback_interface)]
+pub trait NavEventHandler: Send + Sync {
+    /// This callback instruments events that occur when your user navigates to a
+    /// new view. You can add serialized metadata to these events as a byte buffer
+    /// through the [NavOptions] object.
+    fn handle_event(&self, event: NavEvent) -> HandlerResponse;
+}
+
+#[derive(uniffi::Enum, Clone, Debug, PartialEq, Default)]
+pub enum HandlerResponse {
+    #[default]
+    Default,
+    PreventDefault,
+}
+
+#[derive(uniffi::Enum, Clone, Debug, PartialEq)]
+pub enum NavEventType {
+    Push,
+    Replace,
+    Reload,
+    Traverse,
+}
+
+#[derive(uniffi::Record, Clone, Debug, PartialEq)]
+pub struct NavHistoryEntry {
+    /// The target url.
+    pub url: String,
+    /// Unique id for this piece of nav entry state.
+    pub id: HistoryId,
+    /// state passed in by the user, to be passed in to the navigation event callback.
+    pub state: Option<Vec<u8>>,
+}
+
+/// An event emitted when the user navigates between views.
+#[derive(uniffi::Record, Clone, Debug, PartialEq)]
+pub struct NavEvent {
+    pub event: NavEventType,
+    pub same_document: bool,
+    /// The previous location of the page, if there was one
+    pub from: Option<NavHistoryEntry>,
+    /// Destination URL
+    pub to: NavHistoryEntry,
+    /// Additional user provided metadata handed to the event handler.
+    pub info: Option<Vec<u8>>,
```

Comment:
Could this match the name `state` in the `NavHistoryEntry`?

Reply:
`info` and `state` are two different things, the history entry contains persistent state which will always be included whenever it shows up in a navigation event, or when `NavCtx::current` accquires it. `info` is ephemeral and is passed to the event once.

I based this off of the web [navigation api's naming scheme](https://developer.mozilla.org/en-US/docs/Web/API/Navigation/navigate#info) which is honestly not great.

Reply:
`event_specific_info` might be more descriptive.